From a3b0e61397eaf511245aaa41d9b9a818fc38f3e6 Mon Sep 17 00:00:00 2001
From: Zidan Wang <zidan.wang@freescale.com>
Date: Fri, 10 Apr 2015 09:52:36 +0800
Subject: [PATCH 0960/1594] MLK-10611-1 ASoC: fsl-sai: Just one device can
 playback(captrue) when using the same SAI

commit 7981a488c4da440db21f0544b519b44636a0cabb from
git://git.freescale.com/imx/linux-2.6-imx.git

Just one device can playback(captrue) when using the same SAI.

Signed-off-by: Zidan Wang <zidan.wang@freescale.com>
---
 sound/soc/fsl/fsl_sai.c |   13 ++++++++++---
 sound/soc/fsl/fsl_sai.h |    1 +
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/sound/soc/fsl/fsl_sai.c b/sound/soc/fsl/fsl_sai.c
index bcc510a..fc7e68c 100644
--- a/sound/soc/fsl/fsl_sai.c
+++ b/sound/soc/fsl/fsl_sai.c
@@ -548,6 +548,11 @@ static int fsl_sai_startup(struct snd_pcm_substream *substream,
 	struct device *dev = &sai->pdev->dev;
 	int ret;
 
+	if (sai->is_stream_opened[tx])
+		return -EBUSY;
+	else
+		sai->is_stream_opened[tx] = true;
+
 	ret = clk_prepare_enable(sai->bus_clk);
 	if (ret) {
 		dev_err(dev, "failed to enable bus clock: %d\n", ret);
@@ -572,9 +577,11 @@ static void fsl_sai_shutdown(struct snd_pcm_substream *substream,
 	struct fsl_sai *sai = snd_soc_dai_get_drvdata(cpu_dai);
 	bool tx = substream->stream == SNDRV_PCM_STREAM_PLAYBACK;
 
-	regmap_update_bits(sai->regmap, FSL_SAI_xCR3(tx), FSL_SAI_CR3_TRCE, 0);
-
-	clk_disable_unprepare(sai->bus_clk);
+	if (sai->is_stream_opened[tx]) {
+		regmap_update_bits(sai->regmap, FSL_SAI_xCR3(tx), FSL_SAI_CR3_TRCE, 0);
+		clk_disable_unprepare(sai->bus_clk);
+		sai->is_stream_opened[tx] = false;
+	}
 }
 
 static const struct snd_soc_dai_ops fsl_sai_pcm_dai_ops = {
diff --git a/sound/soc/fsl/fsl_sai.h b/sound/soc/fsl/fsl_sai.h
index ffe4cb9..c8ccc1d 100644
--- a/sound/soc/fsl/fsl_sai.h
+++ b/sound/soc/fsl/fsl_sai.h
@@ -140,6 +140,7 @@ struct fsl_sai {
 	bool is_dsp_mode;
 	bool sai_on_imx;
 	bool synchronous[2];
+	bool is_stream_opened[2];
 
 	unsigned int mclk_id;
 	unsigned int mclk_streams;
-- 
1.7.5.4

